program-runtime/src/loaded_programs.rs

```diff
@@ -318,13 +322,51 @@ impl LoadedPrograms {
 
     /// Evicts programs which were used infrequently
     pub fn sort_and_evict(&mut self) {
-        // TODO: Sort programs by their usage_counter
-        // TODO: Truncate the end of the list
+        // Sort programs by their usage_counter wile keeping track on the
```

Comment:
This function could potentially remove cached entry for a program at one of the slots while retaining entry for the same program at a different slot. This can lead to issues. 

For example, if the program was upgraded at a recent slot, and that entry gets evicted, the older program entry will become effective. 

The opposite could also lead to problems, where the older cached entry was still being used, but it got evicted. Now there will not be a way to recreate that program as the program account has been modified with the upgraded program.

@Lichtso what are your thoughts?

Reply:
Agree, the eviction phase should only remove all entries of a Pubkey at once or none of them.
The prune phase is responsible for filtering out stale entries of the same Pubkey.

Reply:
I think evicting programs that only have one entry would be ideal. The programs with more than one entry potentially have the older entry because some older slot still needs it. And, if evicted, those entry may not be regenerated. The calls to `prune()` (on creation of new roots) will eventually remove the old entries. At that point the program will become available for eviction.